src/instrumentation.ts

```diff
@@ -293,66 +293,59 @@ export class Instrumentation extends InstrumentationBase {
+        const span = tracer.startSpan(spanName, {
+          attributes: Instrumentation.dropInvalidAttributes({
+            [SemanticAttributes.MESSAGING_SYSTEM]:
+              BullMQAttributes.MESSAGING_SYSTEM,
+            [SemanticAttributes.MESSAGING_CONSUMER_ID]: workerName,
+            [SemanticAttributes.MESSAGING_MESSAGE_ID]: job.id,
+            [SemanticAttributes.MESSAGING_OPERATION]: "receive",
+            [BullMQAttributes.JOB_NAME]: job.name,
+            [BullMQAttributes.JOB_ATTEMPTS]: job.attemptsMade,
+            [BullMQAttributes.JOB_TIMESTAMP]: job.timestamp,
+            [BullMQAttributes.JOB_DELAY]: job.delay,
+            [BullMQAttributes.JOB_REPEAT_KEY]: job.repeatJobKey,
+            ...Instrumentation.attrMap(BullMQAttributes.JOB_OPTS, job.opts),
+            [BullMQAttributes.QUEUE_NAME]: job.queueName,
+            [BullMQAttributes.WORKER_NAME]: workerName,
+            [BullMQAttributes.WORKER_CONCURRENCY]: this.opts?.concurrency,
+            [BullMQAttributes.WORKER_LOCK_DURATION]: this.opts?.lockDuration,
+            [BullMQAttributes.WORKER_LOCK_RENEW]: this.opts?.lockRenewTime,
+            [BullMQAttributes.WORKER_RATE_LIMIT_MAX]: this.opts?.limiter?.max,
+            [BullMQAttributes.WORKER_RATE_LIMIT_DURATION]:
+              this.opts?.limiter?.duration,
+            // Limit by group keys was removed in bullmq 3.x
+            [BullMQAttributes.WORKER_RATE_LIMIT_GROUP]: (
+              this.opts?.limiter as any
+            )?.groupKey,
+          }),
+          kind: SpanKind.CONSUMER,
+          links: Instrumentation.dropInvalidLinks([
+            {
+              context: trace.getSpanContext(producerContext),
+            },
+          ]),
+        });
 
-        return await context.with(messageContext, async () => {
+        const consumerContext = trace.setSpan(currentContext, span);
+
+        return await context.with(consumerContext, async () => {
           try {
             const result = await original.apply(this, [job, ...rest]);
             return result;
           } catch (e) {
             throw Instrumentation.setError(span, e as Error);
           } finally {
-            if (job.finishedOn)
-              span.setAttribute(
-                BullMQAttributes.JOB_FINISHED_TIMESTAMP,
-                job.finishedOn,
-              );
-            if (job.processedOn)
-              span.setAttribute(
-                BullMQAttributes.JOB_PROCESSED_TIMESTAMP,
-                job.processedOn,
-              );
-            if (job.failedReason)
-              span.setAttribute(
-                BullMQAttributes.JOB_FAILED_REASON,
-                job.failedReason,
-              );
+            span.setAttributes(
+              Instrumentation.dropInvalidAttributes({
+                [BullMQAttributes.JOB_FINISHED_TIMESTAMP]: job.finishedOn,
+                [BullMQAttributes.JOB_PROCESSED_TIMESTAMP]: job.processedOn,
```

Comment:
I've seen these kinds of things stored as events in other libraries. That seems like a good place we can move these to too. We can pick that up as another improvement for this instrumentation library.

Reply:
It still feels counter-intuitive to me to have events on a span whose timestamps do not take place during the duration of said span. That said, I think [the spec agrees with you](https://opentelemetry.io/docs/concepts/signals/traces/#when-to-use-span-events-versus-span-attributes). I'll write an issue about it.